hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieMergeOnReadRDD.scala

```diff
@@ -61,26 +64,31 @@ class HoodieMergeOnReadRDD(@transient sc: SparkContext,
   } else {
     new Properties()
   }
+
+  private val requiredSchema = tableState.requiredStructSchema
+
+  private val requiredFieldPosition = requiredSchema.map(f => tableState.tableStructSchema.fieldIndex(f.name))
+
   override def compute(split: Partition, context: TaskContext): Iterator[InternalRow] = {
     val mergeOnReadPartition = split.asInstanceOf[HoodieMergeOnReadPartition]
     val iter = mergeOnReadPartition.split match {
       case dataFileOnlySplit if dataFileOnlySplit.logPaths.isEmpty =>
-        val rows = read(dataFileOnlySplit.dataFile.get, requiredSchemaFileReader)
-        extractRequiredSchema(rows)
+        val rows = readParquetFile(dataFileOnlySplit.dataFile.get, requiredSchemaFileReader)
+        extractRequiredSchema(rows, requiredSchema, requiredFieldPosition)
```

Comment:
This is actually will try to project already projected data, i think we should be able to clean this up.

Projection is actually done by Spark's Parquet reader in `ParquetFileFormat#buildReaderWithPartitionValues` in both Spark 2.4 and 3.2

Reply:
In some cases, it won't work well when pass the required schema to `ParquetFileFormat#buildReaderWithPartitionValues` and enable `spark.sql.parquet.filterPushdown` in parquet 1.12. This's why i extract projected data after call this method.

Reply:
Can you elaborate what you mean by not working well?

Projecting should be done by Parquet Reader to make sure we're reading only strictly necessary data in, but no more.

Reply:
There's actually HUDI-3396 to make sure this is the case

Reply:
In spark3.2 and parquet 1.12, there is a scene that want to count with a filter. Then read a hudi parquet file by a reader that created by `ParquetFileFormat#buildReaderWithPartitionValues` and pass the filter to it and set `requireSchema` to an empty `StructType`. The result is always 0. But it works in spark3.1.2 and parquet 1.10, if I recall it correctly.

Reply:
Discussed offline, i will pick this up as part of addressing HUDI-3396